File: Data Structures & Algorithms/search-2d-matrix/submission-1.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rowIndex = 0 , right= matrix[0].size()-1 , left = 0;
        while(rowIndex<matrix.size()-1 and target>matrix[rowIndex][right])rowIndex++;

        while(left<=right){
            int mid = (left+right)/2;
            if(matrix[rowIndex][mid]==target){
                return true;
            }else if(matrix[rowIndex][mid]>target){
                right--;
            }else left++;
        }
        return false;



        
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-1.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        ans = 0

        def eating_hours(rate: int) -> int:
            hours = 0
            for pile in piles:
                while pile > 0:
                    pile -= rate
                    hours += 1
            return hours

        left, right = 0, 10**9
        while left <= right:
            mid: int = left + (right - left) // 2
            if eating_hours(mid) <= h:
                right = mid - 1
                ans = int(mid)
            else:
                left = mid + 1

        return ans
